apps/bgu/bgu_generator.cpp

```diff
@@ -126,6 +127,86 @@ Matrix<M, N> solve(Matrix<M, M> A, Matrix<M, N> b, Func compute, Var at, bool sk
     return b;
 };
 
+// Solve Ax = b at each x, y, z exploiting the fact that A is
+// symmetric. Compute the result at the given Func and Var.
+template<int M, int N>
+Matrix<M, N> solve_symmetric(Matrix<M, M> A, Matrix<M, N> b,
+                             Func compute, Var at, bool skip_schedule, Target target) {
+
+    // Put the input matrices in a Func to do sqrt-free Cholesky.
+    Var vi, vj;
+    Func f;
+    f(x, y, z, vi, vj) = undef<float>();
+    for (int i = 0; i < M; i++) {
+        for (int j = 0; j < M; j++) {
+            f(x, y, z, i, j) = A(i, j);
+        }
+        for (int j = 0; j < N; j++) {
+            f(x, y, z, i, j + M) = b(i, j);
+        }
+    }
+
+    for (int j = 0; j < M; j++) {
+        // Normalize the jth column starting at the jth row, storing
+        // the normalization factor on the diagonal. The unnormalized
+        // version stays in the jth row.
+        f(x, y, z, j, j) = fast_inverse(f(x, y, z, j, j));
+        for (int i = j + 1; i < M; i++) {
+            f(x, y, z, i, j) *= f(x, y, z, j, j);
+        }
+
+        // Subtract the outer product of the jth column with its
+        // unnormalized version from the rest of the matrix down and
+        // to the right of it.
+        for (int i = j + 1; i < M; i++) {
+            for (int k = j + 1; k < M; k++) {
+                if (k < i) {
+                    // We already did this one. Exploit symmetry
+                    f(x, y, z, i, k) = f(x, y, z, k, i);
+                } else {
+                    f(x, y, z, i, k) -= f(x, y, z, k, j) * f(x, y, z, j, i);
+                }
+            }
+        }
+    }
+
+    // Back substitute
+    Matrix<M, N> result;
+    for (int k = 0; k < N; k++) {
+        for (int j = 0; j < M; j++) {
```

Comment:
Explain each step?

LDL^T x = b

This is the forward substitution step that first solves for z such that Lz = b.

The first loop just does the forward substitution step without scaling.
The next for loop does the scaling by D^-1.